criu/pidfd.c

```diff
@@ -0,0 +1,286 @@
+}
+
+static struct dead_pidfd *lookup_dead_pidfd(unsigned int ino)
+{
+	struct dead_pidfd *dead;
+	struct hlist_head *chain;
+
+	mutex_lock(dead_pidfd_hash_lock);
+	chain = &dead_pidfd_hash[ino % DEAD_PIDFD_HASH_SIZE];
+	hlist_for_each_entry(dead, chain, hash) {
+		if (dead->ino == ino) {
+			mutex_unlock(dead_pidfd_hash_lock);
+			return dead;
+		}
+	}
+	mutex_unlock(dead_pidfd_hash_lock);
+
+	return NULL;
+}
+
+int is_pidfd_link(char *link)
+{
+	/*
+	* pidfs was introduced in Linux 6.9
+	* before which anonymous-inodes were used
+	*/
+	return is_anon_link_type(link, "[pidfd]");
+}
+
+static void pr_info_pidfd(char *action, PidfdEntry *pidfe)
+{
+	pr_info("%s: id %#08x flags %u NSpid %d ino %u\n",
+		action, pidfe->id, pidfe->flags, pidfe->nspid, pidfe->ino
+	);
+}
+
+static int dump_one_pidfd(int pidfd, u32 id, const struct fd_parms *p)
+{
+	struct pidfd_dump_info pidfd_info = {.pidfe = PIDFD_ENTRY__INIT};
+	FileEntry fe = FILE_ENTRY__INIT;
+
+	if (parse_fdinfo(pidfd, FD_TYPES__PIDFD, &pidfd_info))
+		return -1;
+
+	if (p->flags & PIDFD_THREAD) {
+		pr_err("PIDFD_THREAD flag is currently not supported\n");
+		return -1;
+	}
+
+	/*
+	* Check if the pid pidfd refers to is part of process tree
+	* This ensures the process will exist on restore.
+	*/
+	if (pidfd_info.pid != -1 && !pstree_item_by_real(pidfd_info.pid)) {
+		pr_err("pidfd pid %d is not a part of process tree..\n",
+			pidfd_info.pid);
+		return -1;
+	}
+
+	pidfd_info.pidfe.id = id;
+	pidfd_info.pidfe.flags = (p->flags & ~O_RDWR);
+	pidfd_info.pidfe.fown = (FownEntry *)&p->fown;
+
+	fe.type = FD_TYPES__PIDFD;
+	fe.id = pidfd_info.pidfe.id;
+	fe.pidfd = &pidfd_info.pidfe;
+
+	pr_info_pidfd("Dumping", &pidfd_info.pidfe);
+	return pb_write_one(img_from_set(glob_imgset, CR_FD_FILES), &fe, PB_FILE);
+}
+
+const struct fdtype_ops pidfd_dump_ops = {
+	.type = FD_TYPES__PIDFD,
+	.dump = dump_one_pidfd,
+};
+
+static int pidfd_open(pid_t pid, int flags)
+{
+	return syscall(__NR_pidfd_open, pid, flags);
+}
+
+static int create_tmp_process(void)
+{
+	int tmp_process;
+	tmp_process = fork();
+	if (tmp_process < 0) {
+		pr_perror("Could not fork");
+		return -1;
+	} else if (tmp_process == 0) {
+		while(1)
+			sleep(1);
+	}
+	return tmp_process;
+}
+
+static int free_dead_pidfd(struct dead_pidfd *dead)
+{
+	int status;
+
+	if (kill(dead->pid, SIGKILL) < 0) {
+		pr_perror("Could not kill temporary process with pid: %d",
+		dead->pid);
+		goto err_close;
+	}
+
+	if (waitpid(dead->pid, &status, 0) != dead->pid) {
+		pr_perror("Could not wait on temporary process with pid: %d",
+		dead->pid);
+		goto err_close;
+	}
+
+	if (!WIFSIGNALED(status)) {
+		pr_err("Expected temporary process to be terminated by a signal\n");
+		goto err_close;
+	}
+
+	if (WTERMSIG(status) != SIGKILL) {
+		pr_err("Expected temporary process to be terminated by SIGKILL\n");
+		goto err_close;
+	}
+
+	mutex_lock(dead_pidfd_hash_lock);
+	hlist_del(&dead->hash);
+	mutex_unlock(dead_pidfd_hash_lock);
+	return 0;
+err_close:
```

Comment:
nit: the label name sounds like something is going to be closed... I think it is better to call it just "err".

Reply:
Done!